lib/liquid/strainer_template.rb

```diff
@@ -14,40 +14,67 @@ def initialize(context)
     end
 
     class << self
-      def add_filter(filter)
-        return if include?(filter)
-
-        invokable_non_public_methods = (filter.private_instance_methods + filter.protected_instance_methods).select { |m| invokable?(m) }
-        if invokable_non_public_methods.any?
-          raise MethodOverrideError, "Filter overrides registered public methods as non public: #{invokable_non_public_methods.join(', ')}"
+      def add_filter(mod)
+        filter = if mod.is_a?(Class) && mod.ancestors.include?(Liquid::Filter)
+          mod
+        elsif mod.instance_of?(Module)
+          convert_mod_to_filter(mod)
+        else
+          raise(ArgumentError, "wrong argument type Proc (expected Liquid::Filter)")
         end
 
-        include(filter)
+        instance = filter.new
 
-        filter_methods.merge(filter.public_instance_methods.map(&:to_s))
+        filter.invokable_methods.each do |method|
+          filter_instances[method] = instance
+        end
       end
 
       def invokable?(method)
-        filter_methods.include?(method.to_s)
+        filter_instances.key?(method)
+      end
+
+      def invoke(method, context, *args)
+        instance = filter_instances.fetch(method)
+        instance.context = context
```

Comment:
It looks like `filter_instances` are global, in which case this assignment wouldn't be thread-safe.

Reply:
Good point. As I tried to avoid the `.new` I created this.